### PR TITLE
when forcing a rolling update, emit event with reason

### PR DESCRIPTION
**Pull request by sanchezl**
In order to assist with problem determination, report why a rollout was forced by emitting an event with the reason.

### CODE REVIEW

File: pkg/operator/workload_controller_openshiftapiserver_v311_00.go [hunk 1]
```diff
@@ -52,7 +54,15 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
...
 		if currResult.Changed && resourcesThatForceRedeployment.Has(currResult.File) {
-			forceRollingUpdate = true
+			accessor, err := meta.Accessor(currResult.Result)
+			switch {
+			case err != nil:
+				reasonsForForcedRollingUpdate = append(reasonsForForcedRollingUpdate, fmt.Sprintf("%v modified", currResult.File))
+			case accessor.GetNamespace() == "":
+				reasonsForForcedRollingUpdate = append(reasonsForForcedRollingUpdate, fmt.Sprintf("%v/%v modified", currResult.Result.GetObjectKind().GroupVersionKind().Kind, accessor.GetName()))
+			default:
+				reasonsForForcedRollingUpdate = append(reasonsForForcedRollingUpdate, fmt.Sprintf("%v/%v.%v modified", currResult.Result.GetObjectKind().GroupVersionKind().Kind, accessor.GetName(), accessor.GetNamespace()))
```
Comment (deads2k) [inline pkg/operator/workload_controller_openshiftapiserver_v311_00.go]: `kind.group/name in namespace`

File: pkg/operator/workload_controller_openshiftapiserver_v311_00.go [hunk 2]
```diff
@@ -77,8 +87,28 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
 		errors = append(errors, fmt.Errorf("%q: %v", "client-ca", err))
 	}
 
-	forceRollingUpdate = forceRollingUpdate || operatorConfig.ObjectMeta.Generation != operatorConfig.Status.ObservedGeneration
-	forceRollingUpdate = forceRollingUpdate || configMapModified || etcdModified || clientCAModified || imageImportCAModified
+	if operatorConfig.ObjectMeta.Generation != operatorConfig.Status.ObservedGeneration {
+		reasonsForForcedRollingUpdate = append(reasonsForForcedRollingUpdate, fmt.Sprintf("config generation %q, does not match observed generation %q.", operatorConfig.ObjectMeta.Generation, operatorConfig.Status.ObservedGeneration))
```
Comment (deads2k) [inline pkg/operator/workload_controller_openshiftapiserver_v311_00.go]: "operator config spec generation

File: pkg/operator/workload_controller_openshiftapiserver_v311_00.go [hunk 3]
```diff
@@ -77,8 +87,28 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
...
+	}
+	if configMapModified {
+		reasonsForForcedRollingUpdate = append(reasonsForForcedRollingUpdate, "configmap modified")
```
Comment (deads2k) [inline pkg/operator/workload_controller_openshiftapiserver_v311_00.go]: `configmap/<name> in <namespace>`

File: pkg/operator/workload_controller_openshiftapiserver_v311_00.go [hunk 4]
```diff
@@ -77,8 +87,28 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
...
+	}
+	if etcdModified {
+		reasonsForForcedRollingUpdate = append(reasonsForForcedRollingUpdate, "etcd-certs modified")
```
Comment (deads2k) [inline pkg/operator/workload_controller_openshiftapiserver_v311_00.go]: I want strings that I can copy/paste into `oc get`

File: pkg/operator/workload_controller_openshiftapiserver_v311_00.go [hunk 5]
```diff
@@ -200,53 +229,75 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
...
+	case caChanged:
+		return configMap, true, nil
+	default:
+		return nil, false, nil
 	}
-	return caChanged, nil
 }
 
-func manageOpenShiftAPIServerImageImportCA_v311_00_to_latest(openshiftConfigClient openshiftconfigclientv1.ConfigV1Interface, client coreclientv1.CoreV1Interface, recorder events.Recorder) (bool, error) {
+func manageOpenShiftAPIServerImageImportCA_v311_00_to_latest(openshiftConfigClient openshiftconfigclientv1.ConfigV1Interface, client coreclientv1.CoreV1Interface, recorder events.Recorder) (*corev1.ConfigMap, bool, error) {
```
Comment (deads2k) [inline pkg/operator/workload_controller_openshiftapiserver_v311_00.go]: @sanchezl start a pull on top of something like https://github.com/openshift/cluster-kube-apiserver-operator/pull/185 that moves this into configobservation

File: pkg/operator/workload_controller_openshiftapiserver_v311_00.go [hunk 6]
```diff
@@ -338,3 +389,29 @@ func manageAPIServices_v311_00_to_latest(client apiregistrationv1client.APIServi
...
+	kind := obj.GetObjectKind().GroupVersionKind().Kind
+	version := obj.GetObjectKind().GroupVersionKind().Version
+	group := obj.GetObjectKind().GroupVersionKind().Group
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return "unable to determine resource name"
+	}
+	name := accessor.GetName()
+	namespace := accessor.GetNamespace()
+	var sb strings.Builder
```
Comment (deads2k) [inline pkg/operator/workload_controller_openshiftapiserver_v311_00.go]: we don't use strings.builder

File: pkg/operator/workload_controller_openshiftapiserver_v311_00.go [hunk 7]
```diff
@@ -52,33 +55,59 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
...
 	if err != nil {
 		errors = append(errors, fmt.Errorf("%q: %v", "client-ca", err))
 	}
 
-	forceRollingUpdate = forceRollingUpdate || operatorConfig.ObjectMeta.Generation != operatorConfig.Status.ObservedGeneration
-	forceRollingUpdate = forceRollingUpdate || configMapModified || etcdModified || clientCAModified || imageImportCAModified
+	if operatorConfig.ObjectMeta.Generation != operatorConfig.Status.ObservedGeneration {
+		reasonsForForcedRollingUpdate = append(reasonsForForcedRollingUpdate, fmt.Sprintf("operator config spec generation %q does not match status generation %q", operatorConfig.ObjectMeta.Generation, operatorConfig.Status.ObservedGeneration))
+	}
+	if configMapModified {
```
Comment (deads2k) [inline pkg/operator/workload_controller_openshiftapiserver_v311_00.go]: move these to right where they are being managed.  Otherwise we're going to start forgetting them.

**Comment by deads2k** [comment]
start filling in descriptions.

File: pkg/operator/workload_controller_openshiftapiserver_v311_00.go [hunk 8]
```diff
@@ -52,33 +55,59 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
 		}
 
 		if currResult.Changed && resourcesThatForceRedeployment.Has(currResult.File) {
-			forceRollingUpdate = true
+			if err != nil {
```
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator/workload_controller_openshiftapiserver_v311_00.go]: whoops. fixed.

File: pkg/operator/workload_controller_openshiftapiserver_v311_00.go [hunk 9]
```diff
@@ -52,33 +56,56 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
...
-	forceRollingUpdate = forceRollingUpdate || configMapModified || etcdModified || clientCAModified || imageImportCAModified
+	if operatorConfig.ObjectMeta.Generation != operatorConfig.Status.ObservedGeneration {
+		reasonsForForcedRollingUpdate = append(reasonsForForcedRollingUpdate, fmt.Sprintf("operator config spec generation %q does not match status generation %q", operatorConfig.ObjectMeta.Generation, operatorConfig.Status.ObservedGeneration))
+	}
+
+	forceRollingUpdate := len(reasonsForForcedRollingUpdate) > 0
+
+	if forceRollingUpdate {
+		sort.Sort(sort.StringSlice(reasonsForForcedRollingUpdate))
+		c.eventRecorder.Eventf("RollingUpdateForced", "Rolling update forced because:\n* %s", strings.Join(reasonsForForcedRollingUpdate, "\n* "))
```
Comment (deads2k) [inline pkg/operator/workload_controller_openshiftapiserver_v311_00.go]: nit remove star

**Comment by deads2k** [comment]
/lgtm

rebase

File: pkg/operator/workload_controller_openshiftapiserver_v311_00.go (hunk 5 above)
Comment (sanchezl) [reply pkg/operator/workload_controller_openshiftapiserver_v311_00.go]: Should I wait for https://github.com/openshift/library-go/pull/181 ?

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go [hunk 10]
```diff
@@ -52,22 +56,36 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
...
-	forceRollingUpdate = forceRollingUpdate || configMapModified || imageImportCAModified
+	if operatorConfig.ObjectMeta.Generation != operatorConfig.Status.ObservedGeneration {
+		reasonsForForcedRollingUpdate = append(reasonsForForcedRollingUpdate, fmt.Sprintf("operator config spec generation %q does not match status generation %q", operatorConfig.ObjectMeta.Generation, operatorConfig.Status.ObservedGeneration))
+	}
+
+	forceRollingUpdate := len(reasonsForForcedRollingUpdate) > 0
+
+	if forceRollingUpdate {
+		sort.Sort(sort.StringSlice(reasonsForForcedRollingUpdate))
+		c.eventRecorder.Eventf("RollingUpdateForced", "Rolling update forced because:\n* %s", strings.Join(reasonsForForcedRollingUpdate, "\n"))
```
Comment (sttts) [inline pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: the word "forced" is misleading. Nothing is forced here, but triggered.

We have the `ForceRedeploymentReason` field. Why doesn't this show up here?

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go [hunk 11]
```diff
@@ -311,3 +337,40 @@ func manageAPIServices_v311_00_to_latest(client apiregistrationv1client.APIServi
...
+
+func resourceSelectorForCLI(obj runtime.Object) string {
+	groupVersionKind := obj.GetObjectKind().GroupVersionKind()
+	if len(groupVersionKind.Kind) == 0 {
+		if kinds, _, _ := scheme.Scheme.ObjectKinds(obj); len(kinds) > 0 {
+			groupVersionKind = kinds[0]
+		}
+	}
+	if len(groupVersionKind.Kind) == 0 {
+		if kinds, _, _ := openshiftScheme.ObjectKinds(obj); len(kinds) > 0 {
```
Comment (sttts) [inline pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: why is this necessary?

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go [hunk 12]
```diff
@@ -206,29 +224,37 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
 	return false, nil
 }
 
-func manageOpenShiftAPIServerImageImportCA_v311_00_to_latest(openshiftConfigClient openshiftconfigclientv1.ConfigV1Interface, client coreclientv1.CoreV1Interface, recorder events.Recorder) (bool, error) {
+func manageOpenShiftAPIServerImageImportCA_v311_00_to_latest(openshiftConfigClient openshiftconfigclientv1.ConfigV1Interface, client coreclientv1.CoreV1Interface, recorder events.Recorder) (*corev1.ConfigMap, bool, error) {
```
Comment (sttts) [inline pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: godoc return values.

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go (hunk 10 above)
Comment (deads2k) [reply pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: different operator.

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go (hunk 12 above)
Comment (deads2k) [reply pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: > godoc return values.

@sanchezl update

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go [hunk 13]
```diff
@@ -52,22 +56,36 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
...
 		errors = append(errors, fmt.Errorf("%q: %v", "client-ca", err))
 	}
+	if imageImportCAModified {
+		reasonsForForcedRollingUpdate = append(reasonsForForcedRollingUpdate, "modified: "+resourceSelectorForCLI(imageImportCAModifiedObject))
+	}
 
-	forceRollingUpdate = forceRollingUpdate || operatorConfig.ObjectMeta.Generation != operatorConfig.Status.ObservedGeneration
-	forceRollingUpdate = forceRollingUpdate || configMapModified || imageImportCAModified
+	if operatorConfig.ObjectMeta.Generation != operatorConfig.Status.ObservedGeneration {
+		reasonsForForcedRollingUpdate = append(reasonsForForcedRollingUpdate, fmt.Sprintf("operator config spec generation %q does not match status generation %q", operatorConfig.ObjectMeta.Generation, operatorConfig.Status.ObservedGeneration))
```
Comment (deads2k) [inline pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: `%q` or `%d`

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go [hunk 14]
```diff
@@ -311,3 +337,40 @@ func manageAPIServices_v311_00_to_latest(client apiregistrationv1client.APIServi
...
+	var name string
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		name = "unknown"
+	}
+	name = accessor.GetName()
+	if len(group) > 0 {
+		group = "." + group
+	}
+	return kind + group + "/" + name
```
Comment (mfojtik) [inline pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: @soltysh @sttts i saw this code so many times... last time in event recorder :-)

**Comment by deads2k** [comment]
@sanchezl if you get to it, update.  otherwise this would have helped me yesterday

/lgtm
lgetm
